talk/tools/profiling.tex

```diff
@@ -4,39 +4,228 @@
+$ # (Note that you need to use system-wide sampling for IBS on AMD CPUs)
+$ perf record -a -g -e cycles:pp -- <command>
+
+$ # Sample CPU stack traces once every 10k L1 data cache misses, for 5s:
+$ perf record -a -g -e L1-dcache-load-misses -c 10000 -- sleep 5
+
+$ # Sample CPUs at 100 Hertz, and show top addresses and symbols, live:
+$ perf top -F 100
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Reporting and annotating source code with \mintinline{bash}{perf}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # Standard reporting of perf.data in text UI interface:
+$ perf report
+
+$ # Report by self-time (excluding time spent in callees):
+$ perf report --no-children
+
+$ # Report per source line of code (needs debugging info to work):
+$ perf report --no-children -s srcline
+
+$ # Single inverted (caller-based) call-graph per binary:
+$ perf report --inverted -s comm
+
+$ # Text-based report per library, without call graph:
+$ perf report --stdio -g none -s dso
+
+$ # Hierarchical report for functions taking at least 1% of runtime:
+$ perf report --stdio -g none --hierarchy --percent-limit 1
+
+$ # Disassemble and annotate a symbol (instructions with percentages):
+$ # (Needs debugging information available to show source code as well)
+$ perf annotate <symbol>
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Further information on \mintinline{bash}{perf}}
+  \begin{itemize}
+    \item Official documentation in the Linux repository at
+    \href{https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/perf/Documentation}
+         {linux/tools/perf/Documentation}
+    \item Perf Wiki at \url{https://perf.wiki.kernel.org/}
+    \item Linux \mintinline{bash}{perf} examples by Brendan Gregg
+          \url{https://www.brendangregg.com/linuxperf.html}
+    \item Scripts to visualize profiles as flamegraphs
+          \url{https://github.com/brendangregg/FlameGraph}
+    \item HSF Tools \& Packaging Working Group talk on Indico\\
+          \href{https://indico.cern.ch/event/974382/}
+          {Linux Systems Performance: Tracing, Profiling \& Visualization}
+  \end{itemize}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Intel VTune Profiler}
+  \centering
+  \includegraphics[width=0.75\textwidth]{tools/vtune.png}
+  \begin{itemize}
+    \item Very powerful GUI-based profiler for Intel CPUs and GPUs
+    \item Now free to use with
+      \href{https://www.intel.com/content/www/us/en/developer/tools/oneapi/toolkits.html}{Intel oneAPI Base Toolkit} or
+      \href{https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html}{standalone}
+    \item See the \href{https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/}
+                       {official online documentation} for more information
+  \end{itemize}
 \end{frame}
```

Comment:
Does the picture brings something for people not knowing the tool ? I would maybe replace it with a bullet highlighting the things it can do which perf cannot (if any) and another giving the donwsides

Reply:
Since VTune is a graphical tool, I thought it would be nice to show what it looks like when you open it. You can use the picture to show the types of analyses that VTune is able to do instead of a bullet list, and just tell people when presenting about the extra features it has over perf. For detailed usage information, I'd point people to the online docs. One thing I'd mention while presenting is the Top-Down Microarchitecture Analysis, which is a very good method to find bottlenecks. While perf can also do it, it cannot show you detailed information for each symbol like VTune does, and the annotation of source code by VTune is also a lot easier to use than perf's.

Reply:
I would even like to have more pictures. E.g. I love the microarchitecture analysis with the pipeline visualization. Or how a general hierarchical profile looks like. Or the pane showing contention between threads. Or even better, a live demonstration :)

Reply:
I do not care about picture themselves. I care that if there is a picture, it's understandable, that is that we explain what appears there. In this case, there is a LOT of explanations missing, and I'm not sure we want to include them actually.